datatree/treenode.py

```diff
@@ -243,11 +243,34 @@ def iter_parents(self: Tree) -> Iterator[Tree]:
             yield node
             node = node.parent
 
+    def iter_lineage(self: Tree) -> Iterator[Tree]:
+        """Iterate up the tree, starting from the current node."""
+        from warnings import warn
+
+        warn(
+            "`iter_lineage` has been deprecated, and in the future will raise an error."
+            "Please use `iter_parents` from now on.",
```

Comment:
`iter_parents` can be found above at line 239; the diff is a little bit misleading

- `iter_lineage` became `iter_parents`
- `iter_lineage` now appears as a new patch, and delegates to `iter_parents`

It works in the current situation. Should I replace `other.iter_lineage()` by `other.parents` in https://github.com/xarray-contrib/datatree/blob/66397e89529b2a3ecaa84d4945bd00da12a573a7/datatree/treenode.py#L627 as suggested, or is it fine to leave it as it is, since `iter_lineage` does still exist, delegating to `iter_parents`?

Reply:
I think we don't need `iter_parents/lineage` as well as `parents/lineage`, and we should replace any instances of `iter_lineage` with just the new `parents`.

Reply:
Sorry, this simple renaming becomes longer than expected! My next comment could have been discussed in the issue, but I'm unraveling things as I implement. For next similar issues, I'll inspect `pathlib` more in-depth before implementing. I underestimated the effort to match pathlib behaviour.

I noticed a discrepency with `pathlib` behaviour: in current implementation of `parents`, the node itself is present in the list, which is not the case in pathlib.

Concrete simple example:

```python
In [1]: from pathlib import PurePath

In [2]: p = PurePath('/a/b/c')

In [3]: p.parents
Out[3]: <PurePosixPath.parents>

In [4]: p.parents[0]
Out[4]: PurePosixPath('/a/b')

In [5]: tuple(p.parents)
Out[5]: (PurePosixPath('/a/b'), PurePosixPath('/a'), PurePosixPath('/'))

In [6]: r = PurePath('/')

In [7]: r.parents
Out[7]: <PurePosixPath.parents>

In [8]: tuple(r.parents)
Out[8]: ()

In [9]: p.parent
Out[9]: PurePosixPath('/a/b')

In [10]: r.parent
Out[10]: PurePosixPath('/')
```

We can see that `/a/b/c` is not present in the result, which contradicts `test_parents` where the node itself is present. 

Also, for a root node, an empty sequence is retrieved, not None as currently done (eg with `ancestors`)

Also, the `ancestors` property seems redundant with `parents`, as `ancestors` is just a reversed `parents`. For simplicty, I suggest removing it.

To summarize (there is some _scope creep_, however all properties are related to `parents`), I'd like to do the following:

- `iter_parents` become private: `_iter_parents`, to reuse the existing iterator-related logic for free (and keeping the separation between the Iterator returned by `iter_parents` vs the concrete tuple returned by `parents`)
- `parents` uses `_iter_parents`, but excludes the node itself from its `parents` tuple, to match pathlib behaviour. This would be a breaking change.
- `iter_lineage` now delegates to `parents`. It does not return an iterator anymore, but a tuple, as it is wrapping `parents`. This should not break code using `iter_lineage` as one can iterate over a tuple
- Remove `ancestors` from public-facing API. Keep it internally as as shorthand `_ancestors`, but it is really just a reversed `parents`.
- Add a `parent` property, that is just returning `parents[0]`, like pathlib. Edge case: for a root node, `parent` returns itself.

What do you think? Maybe some changes I proposed are a bit too much. I try to align on pathlib, but maybe following blindly the pathlib philosophy might not always be the most relevant?

Reply:
Hi,

> I agree that excluding self might be neater. As .parents will be an entirely new method, this would not be a breaking change, surely? If would just be a difference in behaviour between .lineage and .parents (which we should note in the warning message).

Yes indeed. I can do the same for `ancestors`. I tried to remove in the internal code all instances of `iter_lineage`, `lineage`, `ancestors`, and replace them by the single pathlib-like new `parents`. All those three methods have been marked for deprecation. They still pass tests for existing user code, but are not used internally anymore, making them more easily removeable in the future. The idea is to have only one pathlib-like method, `parents`, and remove other that can be computed from it.